File: swampinabox/singleserver/health_scripts/showSWAMPHTCondor.php
<?php
function get_condor_submit_node() {
	$command = "condor_status -schedd -af Name";
	exec($command, $output, $returnVar);
	$submit_node = "";
	if (($returnVar == 0) && (! empty($output))) {
		$submit_node = str_replace('"', '', $output[0]);
	}
	return $submit_node;
}

function get_condor_exec_nodes() {
	$command = "condor_status -af Machine -constraint 'SlotType == \"Partitionable\"'";
	exec($command, $output, $returnVar);
	$exec_nodes = [];
	if (($returnVar == 0) && (! empty($output))) {
		for ($i = 0; $i < sizeof($output); $i++) {
			$exec_nodes[$i] = str_replace('"', '', $output[$i]);
		}
	}
	return $exec_nodes;
}

function get_swamp_data_node() {
	// FIXME how should this be obtained in the multi host instance of swamp
	$data_node = gethostname();
	return $data_node;
}

function get_condor_collector_host() {
	// FIXME how should this be obtained in the multi host instance of swamp
	$collector_host = gethostname();
	return $collector_host;
}

$collector_host = get_condor_collector_host();

$global_fields['assessment'] = [
	'Name', 
	'SWAMP_vmu_assessment_vmhostname', 
	'SWAMP_vmu_assessment_status'
];
$global_constraint['assessment'] = "-constraint \"isString(SWAMP_vmu_assessment_status)\"";
$global_fields['viewer'] = [
	'Name',
	'SWAMP_vmu_viewer_vmhostname',
	'SWAMP_vmu_viewer_name',
	'SWAMP_vmu_viewer_state',
	'SWAMP_vmu_viewer_status',
	'SWAMP_vmu_viewer_vmip',
	'SWAMP_vmu_viewer_project',
	'SWAMP_vmu_viewer_instance_uuid',
	'SWAMP_vmu_viewer_apikey',
	'SWAMP_vmu_viewer_url_uuid'
];
$global_constraint['viewer'] = "-constraint \"isString(SWAMP_vmu_viewer_status)\"";

function get_collector_records($collector_host, $title) {
	global $global_fields, $global_constraint;
	$fieldnames = [];
	$crecords = [];
	$fields = $global_fields[$title];
	$constraint = $global_constraint[$title];
	$sortfield = "SWAMP_vmu_" . $title . "_vmhostname";
	$command = "condor_status -pool $collector_host -sort $sortfield -generic -af:V, ";
	foreach ($fields as $field) {
		$command .= ' ' . $field;
	}
	if (! empty($constraint)) {
		$command .= ' ' . $constraint;
	}
	exec($command, $output, $returnVar);
	if (($returnVar == 0) && (! empty($output))) {
		$prefix = "SWAMP_vmu_" . $title . "_";
		$fieldnames[0] = 'execrunuid';
		for ($i = 1; $i < sizeof($fields); $i++) {
			$fieldnames[$i] = str_replace($prefix, '', $fields[$i]);
		}
		for ($i = 0; $i < sizeof($output); $i++) {
			$crecord = [];
			$temp = preg_split("/[\s]+/", $output[$i], sizeof($fieldnames), PREG_SPLIT_NO_EMPTY);
			for ($n = 0; $n < sizeof($fieldnames); $n++) {
				$fieldname = $fieldnames[$n];
				$crecord[$fieldname] = str_replace('"', '',  $temp[$n]);
			}
			$crecords[] = $crecord;
		}
	}
	$ra =  array('fieldnames' => $fieldnames, 'data' => $crecords);
	return $ra;
}

function get_condor_queue() {
	$fieldnames = [];
	$jobs = [];
	$summary = [];
	$command = "condor_q";
	exec($command, $output, $returnVar);
	if (($returnVar == 0) && (! empty($output))) {
		$start = 0;
		for ($i = 0; $i < sizeof($output); $i++) {
			// skip empty lines
			if (empty($output[$i])) {
				continue;
			}
			// collect fieldnames
			if (preg_match('/ID/', $output[$i])) {
				$start = 1;
				$fieldnames = preg_split("/[\s]+/", $output[$i], -1, PREG_SPLIT_NO_EMPTY);
			}
			// collect job summary
			elseif (preg_match('/jobs;/', $output[$i])) {
				$start = 0;
				$temp = preg_split("/[\s]+/", $output[$i], -1, PREG_SPLIT_NO_EMPTY);
				for ($n = 0; $n < sizeof($temp); $n += 2) {
					$name = str_replace('"', '', $temp[$n+1]);
					$name = str_replace(';', '', $name);
					$name = str_replace(',', '', $name);
					$summary[$name] = $temp[$n];
				}
			}
			// collect job records
			elseif ($start == 1) {
				$job = [];
				$temp = preg_split("/[\s]+/", $output[$i], -1, PREG_SPLIT_NO_EMPTY);
				$j = 0;
				for ($n = 0; $n < sizeof($fieldnames); $n++) {
					$fieldname = $fieldnames[$n];
					if ($fieldname == "SUBMITTED") {
						$job[$fieldname] = $temp[$j] . " " . $temp[$j+1];
						$j += 1;
					}
					else {
						$job[$fieldname] = $temp[$j];
					}
					$j += 1;
				}
				$jobs[] = $job; 
			}
		}
	}
	$ra = array('fieldnames' => $fieldnames, 'data' => $jobs, 'summary' => $summary);
	return $ra;
}

function _sort_on_clusterid($commandfield) {
	return function($a, $b) use ($commandfield) {
		$acommand = $a[$commandfield];
		$bcommand = $b[$commandfield];
		$aparts = preg_split("/[\s]+/", $acommand, -1, PREG_SPLIT_NO_EMPTY);
		$bparts = preg_split("/[\s]+/", $bcommand, -1, PREG_SPLIT_NO_EMPTY);
		if (sizeof($aparts) >= 3) {
			// either clusterid procid or clusterid procid [a|v|m]swamp-clusterid-procid
			// sort on clusterid
			list($acid, $apid, $aname) = array_slice($aparts, -3);
			if (is_numeric($aname)) {
				$acid = $apid;
			}
		}
		else {
			// sort on last element
			$temp = array_slice($aparts, -1);
			$acid = array_pop($temp);
		}
		if (sizeof($bparts) >= 3) {
			list($bcid, $bpid, $bname) = array_slice($bparts, -3);
			if (is_numeric($bname)) {
				$bcid = $bpid;
			}
		}
		else {
			$temp = array_slice($bparts, -1);
			$bcid = array_pop($temp);
		}
		if (is_numeric($acid) && is_numeric($bcid)) {
			if ($acid < $bcid) {
				return -1;
			}
			elseif ($acid > $bcid) {
				return 1;
			}
			return 0;
		}
		else {
			return strcmp($acid, $bcid); 
		}
	};
}

function get_swamp_processes() {
	$fieldnames = [];
	$vmuA = [];
	$vmuV = [];
	$vmu = [];
	$java = [];
	$other = [];
	$command = "ps aux | egrep 'PID|vmu_|java' | grep -v grep";
	exec($command, $output, $returnVar);
	if (($returnVar == 0) && (! empty($output))) {
		$commandfield = "";
		for ($i = 0; $i < sizeof($output); $i++) {
			// skip empty lines
			if (empty($output[$i])) {
				continue;
			}
			// collect field names
			if (preg_match('/PID/', $output[$i])) {
				$fieldnames = preg_split("/[\s]+/", $output[$i], -1, PREG_SPLIT_NO_EMPTY);
			}
			else {
				$process = [];
				// collect all of the COMMAND field in the last element of the split
				$temp = preg_split("/[\s]+/", $output[$i], sizeof($fieldnames), PREG_SPLIT_NO_EMPTY);
				for ($n = 0; $n < sizeof($fieldnames); $n++) {
					$fieldname = $fieldnames[$n];
					$process[$fieldname] = $temp[$n];
					$commandfield = $fieldname;
				}
				if (preg_match("/vmu_.*Assessment/", $process[$commandfield])) {
					$vmuA[] = $process;
				}
				elseif (preg_match("/vmu_.*Viewer/", $process[$commandfield])) {
					$vmuV[] = $process;
				}
				elseif (preg_match("/vmu_.*.pl/", $process[$commandfield])) {
					$vmu[] = $process;
				}
				elseif (preg_match("/java/", $process[$commandfield])) {
					$java[] = $process;
				}
				else {
					$other[] = $process;
				}
			}
		}
		// sort vmuA and vmuV on clusterid at end of COMMAND field
		usort($vmuA, _sort_on_clusterid($commandfield));
		usort($vmuV, _sort_on_clusterid($commandfield));
		$processes = array_merge($vmuA, $vmuV, $vmu, $java, $other);
	}
	$ra = array('fieldnames' => $fieldnames, 'data' => $processes);
	return $ra;
}

function get_virtual_machines() {
	$fieldnames = [];
	$machines = [];
	$command = "virsh list --all";
	exec($command, $output, $returnVar);
	if (($returnVar == 0) && (! empty($output))) {
		for ($i = 0; $i < sizeof($output); $i++) {
			// skipt empty lines
			if (empty($output[$i])) {
				continue;
			}
			// skip --- lines
			if (preg_match('/--/', $output[$i])) {
				continue;
			}
			// collect field names
			if (preg_match('/Id/', $output[$i])) {
				$fieldnames = preg_split("/[\s]+/", $output[$i], -1, PREG_SPLIT_NO_EMPTY);
			}
			// collect machines
			else {
				$machine = [];
				$temp = preg_split("/[\s]+/", $output[$i], -1, PREG_SPLIT_NO_EMPTY);
				for ($n = 0; $n < sizeof($fieldnames); $n++) {
					$fieldname = $fieldnames[$n];
					$machine[$fieldname] = $temp[$n];
				}
				$machines[] = $machine;
			}
		}
		// this is a bit of a hack to assume there is a field called Name to sort on
		usort($machines, function($a, $b) { return strcmp($a['Name'], $b['Name']); });
	}
	$ra =  array('fieldnames' => $fieldnames, 'data' => $machines);
	return $ra;
}

function get_submit_job_dirs() {
	// this is a hack because unix/linux ls command does not display field names
	// ls -lrt is presumably guaranteed to produce exactly the following columns
	// log field is added after the ls command by finding the *swamp-<clusterid>-<procid>.log file
	$fieldnames = ['permissions', 'links', 'owner', 'group', 'size', 'modtime', 'dir', 'log'];
	$jobdirs = [];
	$command = "ls -lrt /opt/swamp/run";
	exec($command, $output, $returnVar);
	if (($returnVar == 0) && (! empty($output))) {
		for ($i = 0; $i < sizeof($output); $i++) {
			if (empty($output[$i])) {
				continue;
			}
			if (preg_match('/total/', $output[$i])) {
				continue;
			}
			if (preg_match('/swamp_monitor/', $output[$i])) {
				continue;
			}
			$jobdir = [];
			// modtime parts and dir are in part1[5]
			$part1 = preg_split("/[\s]+/", $output[$i], 6, PREG_SPLIT_NO_EMPTY);
			// modtime = part2[0,1,2]
			// dir = part2[3]
			$part2 = preg_split("/[\s]+/", $part1[5], -1, PREG_SPLIT_NO_EMPTY);
			for ($n = 0; $n < sizeof($fieldnames); $n++) {
				$fieldname = $fieldnames[$n];
				if ($n <= 4) {
					$jobdir[$fieldname] = $part1[$n];
				}
				elseif ($n == 5) {
					$jobdir[$fieldname] = $part2[0] . ' ' . $part2[1] . ' ' . $part2[2];
				}
				else {
					$jobdir[$fieldname] = $part2[3];
				}
			}
			// look in dir for *swamp-<clusterid>-<procid>.log
			$jobdir['log'] = 'n/a';
			if (($dh = opendir('/opt/swamp/run/' . $jobdir['dir'])) !== false) {
				while (false !== ($file = readdir($dh))) {
					if (preg_match('/^[a|m|v]swamp\-\d+\-\d+\.log$/', $file)) {
						$log = str_replace('.log', '', $file);
						if (! $log) {
							$log = 'n/m';
						}
						$jobdir['log'] = $log;
						break;
					}
				}
				closedir($dh);	
			}
			else {
				$jobdir['log'] = 'opendir failed';
			}
			$jobdirs[] = $jobdir;
		}
	}
	$ra = array('fieldnames' => $fieldnames, 'data' => $jobdirs);
	return $ra;
}

function show_result($title, $ra) {
	if (empty($ra)) {
		return;
	}
	print "$title fieldnames: ";
	$fieldnames = $ra['fieldnames'];
	for ($i = 0; $i < sizeof($fieldnames); $i++) {
		print "$fieldnames[$i] ";
	}
	print "\n";
	print "$title data:\n";
	$data = $ra['data'];
	for ($i = 0; $i < sizeof($data); $i++) {
		$datum = $data[$i];
		foreach ($datum as $key => $value) {
			print "$value ";
		}
		print "\n";
	}
	print "\n";
	return $ra;
}

function show_results($ra) {
	$s = show_result("condor queue", $ra['condorqueue']);
	// condor queue summary
	$summary = $s['summary'];
	print "condor queue summary: ";
	foreach ($summary as $key => $value) {
		print "$key $value ";
	}
	print "\n\n";
	show_result("swamp processes", $ra['swampprocesses']);
	show_result("virsh list", $ra['virtualmachines']);
	show_result("submit job directories", $ra['submitjobdirs']);
	show_result("assessment collector", $ra['assessments']);
	show_result("viewer collector", $ra['viewers']);
}

function get_results() {
	global $collector_host;
	// condor queue
	$cq = get_condor_queue();
	// swamp processes
	$sp = get_swamp_processes();
	// virsh list
	$vm = get_virtual_machines();
	// submit job dirs
	$sjd = get_submit_job_dirs();
	// assessment collector	
	$acr = get_collector_records($collector_host, 'assessment');
	// viewer collector	
	$vcr = get_collector_records($collector_host, 'viewer');
	$ra = array(
		'condorqueue' => $cq, 
		'swampprocesses' => $sp, 
		'virtualmachines' => $vm,
		'submitjobdirs' => $sjd,
		'assessments' => $acr,
		'viewers' => $vcr
	);
	return $ra;
}

$ra = get_results();
show_results($ra);

?>

File: swampinabox/singleserver/health_scripts/showSWAMPHTCondor.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Log::Log4perl qw(:easy);
use Getopt::Long;
use HTCondorClassAds qw(
	get_condor_collector_host
	show_collector_records
);
use lib '/opt/swamp/perl5';
use SWAMP::vmu_Support qw(getSwampConfig systemcall);

my $help	= 0;
my $debug	= 0;
my $verbose	= 0;
my $quiet	= 0;
my $html	= 0;

Log::Log4perl->easy_init($verbose ? ($quiet ? $OFF : $DEBUG) : $OFF);
my $hostname = `hostname`; chomp $hostname;
my $username = `whoami`; chomp $username;

sub newline {
	if ($html) {
		return "</br>";
	}
	else {
		return "\n";
	}
}

sub get_swamp_web_node {
	my $web_node = '';
	return $web_node;
}

sub get_swamp_directory_node {
	my $directory_node = '';
	return $directory_node;
}

sub get_swamp_data_node {
	my $data_node = '';
	my $config = getSwampConfig();
	if ($config) {
		$data_node = $config->get('quartermasterHost');
		if ($data_node eq 'localhost') {
			$data_node = $hostname;
		}
	}
	return $data_node;
}

sub get_swamp_condor_node {
	my $condor_node = '';
	return $condor_node;
}

sub get_condor_submit_node {
	my $submit_node = '';
	my $command = "condor_status -schedd -af Name";
	my ($output, $status) = systemcall($command);
	if ($status) {
		print "Error - $command failed - status: $status output: $output", newline() if (! $quiet);
		return $submit_node;
	}
	chomp $output;
	$submit_node = $output;
	return $submit_node;
}

sub get_condor_exec_nodes {
	my @exec_nodes = ();
	my $command = "condor_status -af Machine -constraint 'SlotType == \"Partitionable\"'";
	my ($output, $status) = systemcall($command);
	if ($status) {
		print "Error - $command failed - status: $status output: $output", newline() if (! $quiet);
		return \@exec_nodes;
	}
	@exec_nodes = split "\n", $output;
	return \@exec_nodes;
}

sub show_swamp_processes { my ($host) = @_ ;
	return if (! $host);
	my $command = "ps aux | egrep 'PID|vmu_|java' | grep -v grep";
	if ($host ne $hostname) {
		$command = "ssh $host ps aux | egrep 'vmu_|java' | grep -v grep";
	}
	my ($output, $status) = systemcall($command);
	if ($status) {
		print "Error - $command failed - status: $status output: $output", newline() if (! $quiet);
		return;
	}
	my @lines = split "\n", $output;
	my $fields = "";
	my $vmuA = [];
	my $vmuV = [];
	my $vmu = [];
	my $java = [];
	my $other = [];
	foreach my $line (@lines) {
		if ($line =~ m/PID/) {
			$fields = $line;
		}
		elsif ($line =~ m/vmu_.*Assessment/) {
			push @$vmuA, $line;
		}
		elsif ($line =~ m/vmu_.*Viewer/) {
			push @$vmuV, $line;
		}
		elsif ($line =~ m/vmu_.*.pl/) {
			push @$vmu, $line;
		}
		elsif ($line =~ m/java/) {
			push @$java, $line;
		}
		else {
			push @$other, $line;
		}
	}
	# sort vmuA and vmuV on clusterid at end of COMMAND field
	@$vmuA = sort {
		my @aparts = split ' ', $a; 
		# either clusterid procid or clusterid procid debug
		# 		 -2        -1        -3        -2     -1
		my $acmp = $aparts[-2];
		$acmp = $aparts[-3] if ($aparts[-1] !~ m/^\d+$/);
		my @bparts = split ' ', $b; 
		my $bcmp = $bparts[-2];
		$bcmp = $bparts[-3] if ($bparts[-1] !~ m/^\d+$/);
		if (($acmp =~ m/^\d+$/) && ($bcmp =~ m/^\d+$/)) {
			return $acmp <=> $bcmp;
		}
		else {
			return $acmp cmp $bcmp;
		}
	} @$vmuA;
	@$vmuV = sort {
		my @aparts = split ' ', $a; 
		# either clusterid procid or clusterid procid debug
		# 		 -2        -1        -3        -2     -1
		my $acmp = $aparts[-2];
		$acmp = $aparts[-3] if ($aparts[-1] !~ m/^\d+$/);
		my @bparts = split ' ', $b; 
		my $bcmp = $bparts[-2];
		$bcmp = $bparts[-3] if ($bparts[-1] !~ m/^\d+$/);
		if (($acmp =~ m/^\d+$/) && ($bcmp =~ m/^\d+$/)) {
			return $acmp <=> $bcmp;
		}
		else {
			return $acmp cmp $bcmp;
		}
	} @$vmuV;
	my $header = "vmu_ and java processes on $host";
	# discount fields line in count
	$header .= " T:[" . (scalar(@lines)-1) . "]";
	$header .= " A:[" . (scalar(@$vmuA)) . "]";
	$header .= " V:[" . (scalar(@$vmuV)) . "]";
	$header .= " v:[" . (scalar(@$vmu)) . "]";
	$header .= " j:[" . (scalar(@$java)) . "]";
	$header .= " o:[" . (scalar(@$other)) . "]";
	print $header, newline();
	my $under = "-" x length($header);
	print $under, newline();
	print $fields, newline();
	if (@$vmuA || @$vmuV) {
		print $_, newline() foreach (@$vmuA);
		print $_, newline() foreach (@$vmuV);
	}
	if (@$vmu) {
		print newline();
		print $_, newline() foreach (@$vmu);
	}
	if (@$java) {
		print newline();
		print $_, newline() foreach (@$java);
	}
	if (@$other) {
		print newline();
		print $_, newline() foreach (@$other);
	}
}

sub show_condor_queue { my ($submit_node) = @_ ;
	return if (! $submit_node);
	print "Condor Queue", newline();
	print "------------", newline();
	# my $command = "condor_q -pool $submit_node";
	my $command = "ssh $submit_node condor_q";
	my ($output, $status) = systemcall($command);
	if ($status) {
		print "Error - $command failed - status: $status output: $output", newline() if (! $quiet);
		return;
	}
	if (! $output) {
		print "no condor records", newline() if (! $quiet);
		return;
	}
	chomp $output;
	print $output, newline();
}

sub show_virtual_machines { my ($exec_node) = @_ ;
	return if (! $exec_node);
	my $command = "virsh list --all";
	if ($exec_node ne $hostname) {
		$command = "ssh $exec_node virsh list --all";
	}
	my ($output, $status) = systemcall($command);
	if ($status) {
		print "Error - $command failed - status: $status output: $output", newline() if (! $quiet);
		return;
	}
	my @vms = grep(! m/Id|\-\-/, split "\n", $output);
	# sort vm list on name field with condor clusterid_procid suffix
	@vms = sort {my @aparts = split ' ', $a; my @bparts = split ' ', $b; $aparts[1] cmp $bparts[1];} @vms;
	print $exec_node, " [", scalar(@vms), "]", newline();
	foreach my $vm (@vms) {
		print "  ", $vm, newline();
	}
}

sub show_submit_rundir { my ($submit_node) = @_ ;
	return if (! $submit_node);
	my $command = "ls -lrt /opt/swamp/run";
	if ($submit_node ne $hostname) {
		$command = "ssh $submit_node ls -lrt /opt/swamp/run";
	}
	my ($output, $status) = systemcall($command);
	if ($status) {
		print "Error - $command failed - status: $status output: $output", newline() if (! $quiet);
		return;
	}
	if (! $output) {
		print "no contents in /opt/swamp/run", newline();
		return;
	}
	my @lines = grep(! m/swamp_monitor|total|.bog$/, split "\n", $output);
	my $header = "SWAMP Submit Run Directory:"; 
	print $header, " [" . scalar(@lines) . "]", newline();
	print '-' x length($header), newline();
	foreach my $line (@lines) {
		my $clusterid = '';
		my $dir = (split ' ', $line)[-1];
		my $cfile = (glob("/opt/swamp/run/$dir/ClusterId*"))[0];
		if ($cfile) {
			$clusterid = (split '_', $cfile)[-1];
		}
		print $line, ' ', $clusterid, newline();
	}
	print newline();
}

sub usage {
	print "usage: $0 [-help -debug -verbose -quiet -html -sleep <integer> -environment [dd|dt|it|pd|nn] -condor -collector|[-assess|-viewer] -rundir -libvirt -swamp|[-data|-sub|-exec]]", newline();
	exit;
}

my $environment;
my ($condor, $collector, $assess, $viewer, $rundir, $libvirt, $swamp, $data, $sub, $exec);
my $sleep = 0;

my $result = GetOptions(
	'help'      	=> \$help,
	'debug'     	=> \$debug,
	'verbose'   	=> \$verbose,
	'quiet'			=> \$quiet,
	'html'			=> \$html,
	'sleep=i'		=> \$sleep,
	'environment=s'	=> \$environment,
	'condor'		=> \$condor,
	'collector'		=> \$collector,
	'assess'		=> \$assess,
	'viewer'		=> \$viewer,
	'rundir'		=> \$rundir,
	'libvirt'		=> \$libvirt,
	'swamp'			=> \$swamp,
	'data'			=> \$data,
	'sub'			=> \$sub,
	'exec'			=> \$exec,
);
usage() if ($help || ! $result);    
my $all = ! $condor && ! $collector && ! $assess && ! $viewer && ! $rundir && ! $libvirt && ! $swamp && ! $data && ! $sub && !$exec;

my ($HTCONDOR_COLLECTOR_HOST, $web_node, $directory_node, $data_node, $condor_node, $submit_node, $exec_nodes);
if (! $environment) {
	$HTCONDOR_COLLECTOR_HOST = get_condor_collector_host();
	$web_node = get_swamp_web_node();
	$directory_node = get_swamp_directory_node();
	$data_node = get_swamp_data_node();
	$condor_node = get_swamp_condor_node();
	$submit_node = get_condor_submit_node();
	$exec_nodes = get_condor_exec_nodes();
}
else {
	if ($environment eq 'dd' || $environment eq 'dt' || $environment eq 'it' || $environment eq 'pd') {
		my $external_domain = 'cosalab.org';
		my $internal_domain = 'mirsam.org';
		$HTCONDOR_COLLECTOR_HOST = "swa-csacol-$environment-01.$external_domain";
		$web_node = "swa-csaweb-$environment-01.$internal_domain";
		$directory_node = "swa-dir-$environment-01.$internal_domain";
		$data_node = "swa-csadata-$environment-01.$internal_domain";
		$condor_node = "swa-csacon-$environment-01.$internal_domain";
		$submit_node = "swa-csasub-$environment-01.$internal_domain";
		if ($environment eq 'dd') {
			$exec_nodes = ["swa-exec-$environment-01.$internal_domain"];
		}
		elsif ($environment eq 'dt') {
			$data_node = "swa-csaper-$environment-01.$internal_domain";
			$exec_nodes = [];
			foreach my $nn ('04', '05') {
				push @$exec_nodes, "swa-exec-$environment-$nn.$internal_domain";
			}
		}
		elsif ($environment eq 'it') {
			$exec_nodes = [];
			foreach my $nn ('01', '02') {
				push @$exec_nodes, "swa-exec-$environment-$nn.$internal_domain";
			}
		}
		elsif ($environment eq 'pd') {
			$exec_nodes = [];
			foreach my $nn ('01', '02', '03', '04', '05', '06') {
				push @$exec_nodes, "swa-exec-$environment-$nn.$internal_domain";
			}
		}
	}
	elsif ($environment =~ m/^\d{2}$/) {
		my $host = "swa-exec-dt-$environment";
		$HTCONDOR_COLLECTOR_HOST = $host;
		$web_node = $host;
		$directory_node = $host;
		$data_node = $host;
		$condor_node = $host;
		$submit_node = $host;
		$exec_nodes = [$host];
	}
	else {
		usage();
	}
}

if ($debug) {
	print "hostname: ", $hostname || "", newline();
	print "username: ", $username || "", newline();
	print "environment: ", $environment || "", newline();
	print newline();
	print "collector: ", $HTCONDOR_COLLECTOR_HOST || "", newline();
	print "web: ", $web_node || "", newline();
	print "directory: ", $directory_node || "", newline();
	print "data: ", $data_node || "", newline();
	print "condor: ", $condor_node || "", newline();
	print "submit: ", $submit_node || "", newline();
	print "exec: ", (join ', ', @$exec_nodes) || "", newline();
	print newline();
}

while (1) {
	system("clear") if ($sleep);
	show_condor_queue($submit_node) if ($all || $condor);
	if ($all || $collector || $assess) {
		show_collector_records($HTCONDOR_COLLECTOR_HOST, 'assessment', 1, newline(), $submit_node);
	}
	if ($all || $collector || $viewer) {
		show_collector_records($HTCONDOR_COLLECTOR_HOST, 'viewer', 0, newline(), $submit_node);
	}
	show_submit_rundir($submit_node) if ($all || $rundir);
	if ($all || $libvirt) {
		if ($all) {
			print "Virsh List [", scalar(@$exec_nodes), "]", newline();
			print "----------", newline();
		}
		foreach my $exec_node (@$exec_nodes) {
			next if ($libvirt && ($exec_node ne $hostname));
			show_virtual_machines($exec_node) 
		}
	}
	my $seen = {};
	if ($all || $swamp || $sub) {
		show_swamp_processes($submit_node);
		$seen->{$submit_node} = 1 if ($submit_node);
	}
	if ($all || $swamp || $exec) {
		foreach my $exec_node (@$exec_nodes) {
			next if ($exec && ($exec_node ne $hostname));
			if (! $seen->{$exec_node}) {
				show_swamp_processes($exec_node);
				$seen->{$exec_node} = 1 if ($exec_node);
			}
		}
	}
	if ($all || $swamp || $data) {
		if ($data_node && ! $seen->{$data_node}) {
			show_swamp_processes($data_node) 
		}
	}
	last if (! $sleep);
	sleep $sleep;
}
